tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java

```diff
@@ -0,0 +1,553 @@
+            }
+        }
+
+        return list;
+    }
+
+    private static <T> List<T> append(List<T> a, List<T> b) {
+        if (a.isEmpty()) {
+            return b;
+        }
+        if (b.isEmpty()) {
+            return a;
+        }
+        List<T> result = new ArrayList<>(a.size() + b.size());
+        result.addAll(a);
+        result.addAll(b);
+        return result;
+    }
+
+    private interface IOPredicate<T> {
+        boolean test(T value) throws IOException;
+    }
+
+    private static class RunResult {
+        final java.util.List<FileOperation<?>> ops;
+        final int successes;
+        final Object expected;
+        final Object actual;
+
+        public RunResult(java.util.List<FileOperation<?>> ops, int successes, Object expected, Object actual) {
+            this.ops = ops;
+            this.successes = successes;
+            this.expected = expected;
+            this.actual = actual;
+        }
+
+        public boolean ok() {
+            return Objects.equals(expected, actual);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            RunResult runResult = (RunResult) o;
+            return successes == runResult.successes &&
+                    Objects.equals(expected, runResult.expected) &&
+                    Objects.equals(actual, runResult.actual);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(successes, expected, actual);
+        }
+
+        @Override
+        public String toString() {
+            return "RunResult{" +
+                    "opsExecuted=" + successes +
+                    ", expected=" + expected +
+                    ", actual=" + actual +
+                    '}';
+        }
+    }
+
+    private RunResult checkOps(java.util.List<FileOperation<?>> ops) throws IOException {
+
+		final File tmpFile1 = File.createTempFile("tmp", ".bin");
+		tmpFile1.deleteOnExit();
+		final File tmpFile2 = File.createTempFile("tmp", ".bin");
+		tmpFile2.deleteOnExit();
+
+        try (RandomAccessFile f1 = new RandomAccessFile(tmpFile1, "rw");
+             BufferedRandomAccessFile f2 = new BufferedRandomAccessFile(tmpFile2, "rw")) {
+
+            for (int i = 0; i < ops.size(); ++i) {
+                FileOperation<?> op = ops.get(i);
+                Object a = op.execute(f1);
+                Object b;
+                try {
+                    b = op.execute(f2);
+                } catch (Exception | AssertionError e) {
+                    return new RunResult(ops, i, null, e.getClass());
+                }
+                if (!Objects.equals(a, b)) {
+                    return new RunResult(ops, i, a, b);
+                }
+            }
+
+        }
+
+        return new RunResult(ops, ops.size(), null, null);
+    }
+
+    private boolean legal(java.util.List<FileOperation<?>> ops) {
```

Comment:
minor, but it struck me that your construction of `FileOperation` here is pretty clever, to where it seems like like could instead be the `simulateIfLegal` method on `ListOfOperations` class that implements `FileOperation`.  Seems like it has a valid `execute` and `toString`.

Reply:
Indeed! But there is a fair amount of code that cares about the list structure, so I don't think there's much benefit in implementing `SequencedOperations implements FileOperation`, even if it would be very elegant. :)